### ISSUE TITLE
Unsafe shenigans in constants can result in missing errors or ICE

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#49296]
```rust
#![feature(const_fn)]

const unsafe fn transmute<T: Copy, U: Copy>(t: T) -> U {
    union Transmute<T: Copy, U: Copy> {
        from: T,
        to: U,
    }

    Transmute { from: t }.to
}

const fn wat(x: u64) -> &'static u64 {
    unsafe { transmute(&x) }
}
const X: u64 = *wat(42);

fn main() {
    println!("{}", X);
}
```

produces

```
error[E0080]: constant evaluation error
  --> src/main.rs:18:20
   |
18 |     println!("{}", X);
   |                    ^ referenced constant has errors
```

without any additional info. If the value is not dereferenced, but instead fed to llvm, we get an ICE in trans:

```rust
#![feature(const_fn)]

const unsafe fn transmute<T: Copy, U: Copy>(t: T) -> U {
    union Transmute<T: Copy, U: Copy> {
        from: T,
        to: U,
    }

    Transmute { from: t }.to
}

const fn wat(x: u64) -> u64 {
    unsafe { transmute(&x) }
}
const X: u64 = wat(42);

fn main() {
    println!("{}", X);
}
```

results in

```
error: internal compiler error: librustc_mir/monomorphize/collector.rs:1131: alloc id without corresponding allocation: 39
```

Even though it should be reporting a dangling pointer const eval error

**Comment by memoryruins** [comment]
Both samples now require ``#![feature(const_fn_union)]``

The first reports a dereferenced dangling pointer
```rust
error: this constant cannot be used
  --> src/main.rs:16:1
   |
16 | const X: u64 = *wat(42);
   | ^^^^^^^^^^^^^^^--------^
   |                |
   |                dangling pointer was dereferenced
```

and the second sample no longer triggers an ICE

```rust
error[E0080]: this constant likely exhibits undefined behavior
  --> src/main.rs:16:1
   |
16 | const X: u64 = wat(42);
   | ^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered a pointer, but expected initialized plain bits
   |
   = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rust compiler repository if you believe it should not be considered undefined behavior

error: aborting due to previous error

```
``rustc 1.31.0-nightly (f99911a4a 2018-10-23)``

**Comment by oli-obk** [comment]
Thanks for checking up on this!

So all we need is a test for the first example, because I don't think we have a regression test for that error. The second one definitely has a bunch of tests.
